### PR TITLE
Add int8 dynamic quantization support

### LINKED ISSUE #7: Add quantization to run the computation in int8
If part (or all) of our model was in int8, then it would have a much lower memory footprint, would require a smaller model to download, and would run faster.

Quantization to int4 is unlikely to be useful, because CPUs that have instructions to handle int4 are very rare. It's unlikely also that people have written kernels to run int4 computation on cpu. So we can stick to int8. Some parts of the model might handle quantization better than others, so it's possible we have to run some parts of the model in int8 and some other in bfloat16 or even float32.

### PR DESCRIPTION
**Pull request by nabil-tazi**
# Add int8 dynamic quantization
 
Closes #7
 
## Summary
 
This PR implements mixed-precision int8 dynamic quantization for pocket-tts, as requested in issue #7. It adds a `quantize=True` flag to `TTSModel.load_model()` that reduces runtime memory usage by ~48% and improves inference speed by ~27% on x86 and up to 23% on ARM, with no measurable impact on audio quality.
 
The implementation uses `torch.ao.quantization.quantize_dynamic` applied selectively to specific layer groups, leaving precision-sensitive components in float32 - exactly the mixed-precision approach anticipated in the issue.
 
## Usage
 
```python
# Default - recommended config (attention + ffn quantized)
tts_model = TTSModel.load_model(quantize=True)
 
# Advanced - override which layer groups to quantize
tts_model = TTSModel.load_model(
    quantize=True,
    quantize_groups={"attention", "ffn", "flow_net"}
)
```
 
CLI:
```bash
pocket-tts generate --text "Hello world" --quantize
```
 
## How it works
 
The model has three quantizable layer groups in the FlowLM transformer:
 
- `attention` - Q/K/V/output projections in each transformer layer
- `ffn` - feedforward up/gate/down projections in each transformer layer  
- `flow_net` - the MLP flow matching sampler
 
The Mimi VAE decoder (convolutional) is unaffected - `quantize_dynamic` only targets `nn.Linear` layers.
 
The default config (`quantize=True`) quantizes `attention` + `ffn`, leaving `flow_net` in float32. Users can override this via `quantize_groups` to target any combination.
 
## Results
 
Seven configurations were evaluated across x86 (FBGEMM backend) and ARM (Apple Silicon M4), each run 10 times. Results differ by backend, see the torchao support section below.
 
RTS = real-time speed factor (higher is better).
 
### torch.ao backend (default, works on torch>=2.5)
 
Uses `torch.ao.quantization.quantize_dynamic` with `dtype=torch.qint8`. ARM measured on Apple Silicon M4 (QNNPACK).
 
| Config | Runtime memory | x86 RTS | ARM RTS | WER delta |
|---|---|---|---|---|
| **baseline** | 450 MB | 3.17x ±0.05 | 6.40x ±0.34 | - |
| flow_net | 422 MB (−6%) | 3.33x (+5%) ±0.03 | 6.29x (−2%) ±0.06 | −0.034 ±0.034 |
| attention | 378 MB (−16%) | 3.53x (+11%) ±0.02 | 6.20x (−3%) ±0.09 | −0.021 ±0.027 |
| flow_net_attention | 350 MB (−22%) | 3.69x (+16%) ±0.03 | 6.07x (−5%) ±0.25 | −0.028 ±0.034 |
| ffn | 306 MB (−32%) | 3.92x (+24%) ±0.03 | 5.97x (−7%) ±0.15 | +0.002 ±0.028 |
| ffn_flow_net | 278 MB (−38%) | 3.89x (+23%) ±0.04 | 5.54x (−13%) ±0.06 | +0.019 ±0.024 |
| **attention_ffn** *(default)* | **234 MB (−48%)** | **4.04x (+27%) ±0.03** | **5.36x (−16%) ±0.04** | **−0.022 ±0.032** |
| all | 206 MB (−54%) | 4.01x (+26%) ±0.03 | 4.61x (−28%) ±0.23 | −0.015 ±0.026 |
 
x86 is consistently faster across all configs. ARM shows a penalty on configs containing FFN quantization due to per-operation dequantization overhead in QNNPACK at batch size 1.
 
### torchao backend (pip install pocket-tts[quantize], requires torch>=2.10)
 
Uses `torchao.quantization.quantize_` with `Int8DynamicActivationInt8WeightConfig`. ARM measured on Apple Silicon M4.
 
| Config | Runtime memory | ARM RTS | vs baseline |
|---|---|---|---|
| **baseline** | 450 MB | 6.33x ±0.69 | - |
| flow_net | 422 MB (−6%) | 5.91x ±0.33 | −7% |
| attention | 378 MB (−16%) | 6.74x ±0.60 | +6% |
| flow_net_attention | 350 MB (−22%) | 5.56x ±0.42 | −12% |
| ffn | 306 MB (−32%) | 6.64x ±0.75 | +5% |
| ffn_flow_net | 278 MB (−38%) | 6.21x ±0.92 | −2% |
| **attention_ffn** *(default)* | **234 MB (−48%)** | **7.76x ±0.33** | **+23%** |
| all | 206 MB (−54%) | 6.81x ±0.35 | +8% |
 
With torchao, `attention_ffn` is faster than baseline on ARM (+23%) while also achieving the best size reduction (−63%). The torchao path outperforms the torch.ao path on ARM across all configs.

edit: on x86, the results with torchao are just as good as with torch.ao.
 
## Quality evaluation
 
**Qualitative:** All 8 voices were evaluated across multiple text samples at every configuration. Listening tests showed no noticeable difference in naturalness, clarity, or artefacts compared to the float32 baseline.
 
**Quantitative:** Word Error Rate was measured using Whisper-large-v3 transcription (the same model used in the original pocket-tts evaluation) across all voices and configurations. WER delta for the default `attention_ffn` config is −0.022 ±0.032 - the ± range crosses zero, meaning the delta is indistinguishable from measurement noise. This holds for every configuration tested.
 
A note on quality metrics for autoregressive TTS: SNR and PESQ are not valid quality metrics for this class of model. Because generation is stochastic and autoregressive, two independent runs produce uncorrelated waveforms even at temperature=0 due to quantization noise shifting argmax decisions at individual steps. Comparing them waveform-to-waveform yields deeply negative SNR regardless of perceptual quality - this is sequence divergence, not degradation. WER, which operates on meaning rather than waveform alignment, is the correct metric here.
 
This is relevant context for [the finding in this thread](https://github.com/kyutai-labs/pocket-tts/issues/7) reporting SNR < 0 dB as evidence of quality degradation - that result most likely reflects a misaligned waveform comparison rather than actual audio quality loss.
 
## Why attention_ffn as default
 
`attention_ffn` was chosen as the default for both backends:
 
- **torch.ao path:** best x86 speedup (+27%), meaningful memory reduction (−48%). The ARM penalty (−16%) is acceptable: the model still runs at 5.36x real-time, well above the real-time threshold.
- **torchao path:** best overall result: faster than baseline on ARM (+23%), best size reduction (−63%), x86 data pending.
 
Users who want ARM-neutral behaviour on the torch.ao path can use `quantize_groups={"attention", "flow_net"}` (`flow_net_attention` config, −22% memory, +16% x86, −5% ARM).
 
## Platform behaviour
 
Speed gains are backend-dependent:
 
- **torch.ao (default):** x86 is faster across all configs via genuine int8 GEMM (FBGEMM). ARM shows a penalty on configs containing FFN quantization due to per-operation dequantization overhead in QNNPACK at batch size 1. Attention-only and flow_net-only configs remain near-baseline on ARM.
- **torchao (`pocket-tts[quantize]`):** uses `Int8DynamicActivationInt8WeightConfig` which delivers genuine speedups on ARM Apple Silicon, outperforming the torch.ao path on all configs. Requires torch>=2.10 and torchao>=0.16.
 
Additionally, QNNPACK currently ignores the `reduce_range` flag with a deprecation warning - this has no impact on audio quality or benchmark validity, but ARM results may shift slightly when a future PyTorch release corrects this behaviour.
 
## torchao support
 
The quantization backend is chosen automatically at runtime:
 
- If `torchao` is installed with working C++ extensions (requires `torch>=2.10.0` + `torchao>=0.16.0`), the modern `torchao` path is used.
- Otherwise, the code falls back to `torch.ao.quantization.quantize_dynamic`, which works on any torch version supported by pocket-tts (2.5+).
 
`torchao` is an optional dependency and is not installed by default. Users who want the modern path can install it explicitly:
 
```bash
pip install "pocket-tts[quantize]"
```
 
This avoids forcing a torch version upgrade on users who simply want to `pip install pocket-tts`. The `torch.ao` fallback path is deprecated in PyTorch 2.9 and scheduled for removal in 2.10 - users on older torch versions should be aware it will eventually stop working, but the `torchao` path will remain fully supported going forward.
 
## Files changed
 
- `pocket_tts/quantization.py` - new: `apply_dynamic_int8()`, layer group registry, `RECOMMENDED_CONFIG`
- `pocket_tts/models/tts_model.py` - `load_model()` gains `quantize` and `quantize_groups` parameters
- `pocket_tts/main.py` - `--quantize` flag added to `generate` and `serve` commands
- `pyproject.toml` - adds `quantize` optional dependency group (`torchao>=0.16.0`)
- `tests/test_quantization.py` - new: correctness tests (no NaN, not silent, shape consistency, no crashes across configs)
- `scripts/evaluate_quantization.py` - new: evaluation harness used to produce the results above
- `docs/quantization.md` - new: usage guide and platform notes

### CODE REVIEW

File: pocket_tts/quantization.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    quantized_count = 0
+
+    if "flow_net" in quantize_groups:
+        _quantize_module_torchao(flow_lm.flow_net)
+        quantized_count += sum(
+            1 for _, m in flow_lm.flow_net.named_modules() if isinstance(m, nn.Linear)
+        )
+        logger.info("Quantized flow_net (MLP sampler)")
+
+    if "attention" in quantize_groups or "ffn" in quantize_groups:
```
[the remark this answers is not in the document]
Comment (nabil-tazi) [reply pocket_tts/quantization.py]: done

File: pocket_tts/quantization.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    quantized_count = 0
+
+    if "flow_net" in quantize_groups:
+        quantize_dynamic(flow_lm.flow_net, {nn.Linear}, dtype=torch.qint8, inplace=True)
+        quantized_count += sum(
+            1 for _, m in flow_lm.flow_net.named_modules() if isinstance(m, nn.Linear)
+        )
+        logger.info("Quantized flow_net (MLP sampler)")
+
+    if "attention" in quantize_groups or "ffn" in quantize_groups:
```
[the remark this answers is not in the document]
Comment (nabil-tazi) [reply pocket_tts/quantization.py]: done

**Comment by nabil-tazi** [comment]
Thanks for the quick review! I made the requested updates

**Comment by gabrieldemarmiesse** [comment]
It works really great on my laptop (WSL2, x86) it went from 3x real-time to 4x real-time :)
